### ISSUE TITLE
execute() failing with 403 error

### ISSUE THREAD

**Issue by txase** [hjdhjd/myq#4]
**Describe The Problem:**
When I invoke `myQ.execute()` I get a 403 error.

**To Reproduce:**
Code:

```JS
const { myQApi } = require('@hjdhjd/myq');

const myQ = new myQApi('<email>', '<password>');

if (!await myQ.refreshDevices()) {
  console.log('Failed to get devices from MyQ API');
  return;
}

const device = myQ.getDevice('<garage door device ID>');

if (!await myQ.execute(device, 'close')) {
  console.log('Failed to execute command');
  return;
}
```

**Logs:**
```
2021-09-18T23:03:57.072Z	INFO	myQ API: Successfully connected to the myQ API.
2021-09-18T23:03:57.827Z	INFO	myQ API: Discovered device family garagedoor: Garage Door (serial number: *****, gateway: *****).
2021-09-18T23:03:57.832Z	INFO	myQ API: Discovered device family gateway: Garage Door [Liftmaster WiFi GDO AC] (serial number: *****).
2021-09-18T23:03:58.125Z	ERROR	myQ API: https://account-devices-gdo.myq-cloud.com/api/v5.2/Accounts/*****/door_openers/*****/close Error: 403 Forbidden
2021-09-18T23:03:58.125Z	ERROR	myQ API: Unable to send the command to myQ servers. Acquiring a new access token.
2021-09-18T23:03:58.125Z	INFO	Failed to execute command
```

I can verify the masked out values above appear correct for my device.

**Screenshots:**

N/A

**Comment by txase** [comment]
Ahh, this occurred because the garage door was already in the closed state. That's a terrible way for the MyQ service to behave :(. Does 403 always mean that the device is already in the asked for state? If so, we should look for that and still return `true`.

**Comment by txase** [comment]
Maybe the `execute()` command should check the current device state, and if it matches the desired state it should return `true` and not issue a request? I'm building this logic into my app, and imagine most folks would need similar logic, too. The device state is already passed into the request.

If you think this is the right approach, I would be happy to create a PR. Let me know, or feel free to close this if you don't think that makes sense.

**Comment by hjdhjd** [comment]
No, we shouldn’t. Knowledge of the current state when a given command is executed is never guaranteed in myQ. It only represents a moment in time. It’s incumbent upon you to manage that state. The reason for this is twofold:

1. The command to execute will then be dependent on executing a refresh of the state from the myQ API, which can return quickly or not so quickly…thereby incurring a speed penalty to executing a command that you would rightly expect to return pretty quickly.
2. The way the myQ API works, this is a built in race condition. It is possible to believe the state is open (due to a recent refresh), for instance, but by the time the command executed, it may have been changed manually at the garage door, or through the native app, or a myriad of other ways.

The result is always the same: checking state before executing the command guarantees nothing, and at best minimizes a possible race condition at the expense of speed, performance, and an additional API call.

**Comment by txase** [comment]
I agree there's a race condition. But since that race condition is built into the service I might argue that a library for the service should do its best to meet human expectations in "normal circumstances" given the limitations. There could be a "force" option to force the API to make the call even when the cached device state is already as requested.

That said, an alternative approach might be to simply return `true` if the response is 403, signaling the door is already in the requested state. Do you know if 403 Is only ever returned under this condition? This would at least address the race condition aspect if true.

**Comment by hjdhjd** [comment]
There's no guarantee of what will and won't generate a given error - it's an undocumented API after all. the myQ team does have a habit of occasionally changing the types of errors generated when an event fails.

Remember, this API implementation is well-tested and has been in use for quite a long time. I'm reluctant to make any significant change that involves making assumptions about a user might or might not want to see. What you expect isn't what I expect. :smile: 

This library isn't here to save you from yourself or from the API...it's intended to provide good access to the underlying API and removing most of the complexity in interacting with it. It isn't a substitute for preserving state, nor saving you from getting an error when you try a particular command. If you want those higher-order capabilities tailored to your use case, you can certainly build a wrapper around the requisite function and call that instead. They don't belong in the library itself.

TL;DR: nope. :smile:
